### ISSUE TITLE
Conflicting Type definitions - Typescript vs Webpack Module Resolution

### ISSUE THREAD

**Issue by IanEdington** [TypeStrong/fork-ts-checker-webpack-plugin#116]
I'm having an issue that I believe must be related to the difference in how Webpack and Typescript resolve moduals.

When I use this plugin I get a number of conflicting type definitions.

I'm posting here because I've scoured the internet and haven't found a solution to this.

```
ERROR in project/node_modules/@types/node/index.d.ts
(103,13): Cannot redeclare block-scoped variable 'console'.

ERROR in project/node_modules/@types/node/index.d.ts
(148,13): Duplicate identifier 'require'.

ERROR in project/node_modules/@types/node/index.d.ts
(2381,15): Duplicate identifier 'URL'.

ERROR in project/node_modules/@types/node/index.d.ts
(2399,15): Duplicate identifier 'URLSearchParams'.

ERROR in project/node_modules/@types/node/index.d.ts
(2417,14): Cannot export 'URL'. Only local declarations can be exported from a module.

ERROR in project/node_modules/@types/node/index.d.ts
(2417,19): Cannot export 'URLSearchParams'. Only local declarations can be exported from a module.

ERROR in project/node_modules/@types/react-native/index.d.ts
(8723,14): Duplicate identifier 'require'.

ERROR in project/node_modules/@types/react-native/index.d.ts
(8741,11): Cannot redeclare block-scoped variable 'console'.

ERROR in project/node_modules/@types/react-native/index.d.ts
(8749,18): Subsequent property declarations must have the same type.  Property 'geolocation' must be of type 'Geolocation', but here has type 'GeolocationStatic'.

ERROR in project/node_modules/@types/react-native/index.d.ts
(8752,11): Cannot redeclare block-scoped variable 'navigator'.

ERROR in project/node_modules/@types/webpack-env/index.d.ts
(203,13): Duplicate identifier 'require'.

ERROR in project/node_modules/rxjs/scheduler/VirtualTimeScheduler.d.ts
(24,15): Property 'work' in type 'VirtualAction<T>' is not assignable to the same property in base type 'AsyncAction<T>'.
  Type '(this: VirtualAction<T>, state?: T | undefined) => void' is not assignable to type '(this: AsyncAction<T>, state?: T | undefined) => void'.
    The 'this' types of each signature are incompatible.
      Type 'AsyncAction<T>' is not assignable to type 'VirtualAction<T>'.
        Property 'index' is missing in type 'AsyncAction<T>'.

ERROR in project/node_modules/typescript/lib/lib.dom.d.ts
(13389,11): Duplicate identifier 'URL'.

ERROR in project/node_modules/typescript/lib/lib.dom.d.ts
(13405,13): Duplicate identifier 'URL'.

ERROR in project/node_modules/typescript/lib/lib.dom.d.ts
(13412,11): Duplicate identifier 'URLSearchParams'.

ERROR in project/node_modules/typescript/lib/lib.dom.d.ts
(13439,13): Duplicate identifier 'URLSearchParams'.

ERROR in project/node_modules/typescript/lib/lib.dom.d.ts
(15720,13): Duplicate identifier 'URL'.

ERROR in project/node_modules/typescript/lib/lib.dom.d.ts
(15721,13): Duplicate identifier 'URLSearchParams'.

ERROR in project/node_modules/typescript/lib/lib.dom.d.ts
(15747,13): Cannot redeclare block-scoped variable 'navigator'.

ERROR in project/node_modules/typescript/lib/lib.dom.d.ts
(15920,13): Cannot redeclare block-scoped variable 'console'.
```

Here's the diff of my webpack config:
```
───────────────────────────
modified: config/webpack.config.dev.js
───────────────────────────
@ webpack.config.dev.js:8 @ const webpack = require('webpack');
const CaseSensitivePathsPlugin = require('case-sensitive-paths-webpack-plugin');
const WatchMissingNodeModulesPlugin = require('react-dev-utils/WatchMissingNodeModulesPlugin');
const TsconfigPathsPlugin = require('tsconfig-paths-webpack-plugin');
+const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');
const getClientEnvironment = require('./env');
const paths = require('./paths');

@ webpack.config.dev.js:129 @ module.exports = {
            include: paths.appSrc,
            loaders: [
              require.resolve('babel-loader'),
-              require.resolve('ts-loader'),
+              {
+                loader: 'ts-loader',
+                options: {
+                  transpileOnly: true,
+                },
+              },
            ],
          },
          // Process JS with Babel.
@ webpack.config.dev.js:264 @ module.exports = {
    // https://github.com/jmblog/how-to-optimize-momentjs-with-webpack
    // You can remove this if you don't use Moment.js:
    new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/),
+    new ForkTsCheckerWebpackPlugin({
+      tsconfig: paths.appTsConfig,
+    }),
  ],
  // Some libraries import Node modules but don't use them in the browser.
  // Tell Webpack to provide empty mocks for them so importing them works.
```

**Comment by johnnyreilly** [comment]
Why don't you take a working example and try and plug your code into that? https://github.com/TypeStrong/ts-loader/tree/master/examples/fork-ts-checker

Might make things easier for you.

**Comment by IanEdington** [comment]
Thanks for the quick reply.

I've taken a look a that as well as read the documentation. It looks like my issue is with this: 
https://github.com/Realytics/fork-ts-checker-webpack-plugin#modules-resolution
The Webpack module resolution doesn't import the types that are conflicting however TypeScript does.

@johnnyreilly  Is there something in `examples` that I'm missing that resolves these conflicting Type definitions?

**Comment by johnnyreilly** [comment]
I'm going to guess that the problem either lies in your `tsconfig.json` or in your usage of the `tsconfig-paths-webpack-plugin` (I don't use it myself and so if that is the issue you might want to try asking on that project.)

Does your `tsconfig.json`  include all the files? Might be worth explicitly ensuring that it does and see where that leads.

**Comment by johnnyreilly** [comment]
FWIW I don't supply the `tsconfig.json` to the fork-ts-checker-webpack-plugin myself. Don't know if that's relevant

**Comment by naiyt** [comment]
I was encountering this as well and some some people saying it's possibly to mismatched type definition packages. A `rm yarn.lock && rm -rf node_modules && yarn` fixed it for me.

Of course nuking your lockfile is always scary, because you might end up with a bunch of upgraded packages that don't work with your setup. If that's the case you might have to do some manual surgery on them.

This was the main issue I was reading that led me down that path: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18484

EDIT: actually, scratch that. After the upgrade webpack wasn't running the `fork-ts-checker-webpack-plugin` at all, so I wasn't getting any type checking, which is why the errors weren't showing up. :(

**Comment by qdot** [comment]
This is a problem with node and typescript libraries colliding. See:

https://github.com/Microsoft/TypeScript/issues/23769

Quick fix: Back down to `@types/node` v9.6.7 until v10 can be fixed.

**Comment by IanEdington** [comment]
Thanks everyone!

@qdot's solution worked for me.
